### PR TITLE
[Merged by Bors] - chore: forward-port backports

### PR DESCRIPTION
**Pull request by kim-em**
* `data.mv_polynomial.basic`, `data.mv_polynomial.funext`: leanprover-community/mathlib#18839
* `category_theory.limits.preserves.finite`, `category_theory.preadditive.projective`:  leanprover-community/mathlib#18890
* `category_theory.abelian.basic`, `category_theory.abelian.opposite`:  leanprover-community/mathlib#18740
* `topology.category.Top.limits.basic`: leanprover-community/mathlib#18871. Note that this does not show a useful diff on the dashboard pages as file splits aren't tracked well by git.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)
* [`data.mv_polynomial.basic`@`2d5739b61641ee4e7e53eca5688a08f66f2e6a60`..`0b89934139d3be96f9dab477f10c20f9f93da580`](https://leanprover-community.github.io/mathlib-port-status/file/data/mv_polynomial/basic?range=2d5739b61641ee4e7e53eca5688a08f66f2e6a60..0b89934139d3be96f9dab477f10c20f9f93da580)
* [`category_theory.limits.preserves.finite`@`024a4231815538ac739f52d08dd20a55da0d6b23`..`3974a774a707e2e06046a14c0eaef4654584fada`](https://leanprover-community.github.io/mathlib-port-status/file/category_theory/limits/preserves/finite?range=024a4231815538ac739f52d08dd20a55da0d6b23..3974a774a707e2e06046a14c0eaef4654584fada)
* [`category_theory.abelian.basic`@`8c75ef3517d4106e89fe524e6281d0b0545f47fc`..`a5ff45a1c92c278b03b52459a620cfd9c49ebc80`](https://leanprover-community.github.io/mathlib-port-status/file/category_theory/abelian/basic?range=8c75ef3517d4106e89fe524e6281d0b0545f47fc..a5ff45a1c92c278b03b52459a620cfd9c49ebc80)
* [`topology.category.Top.limits.basic`@`8195826f5c428fc283510bc67303dd4472d78498`..`178a32653e369dce2da68dc6b2694e385d484ef1`](https://leanprover-community.github.io/mathlib-port-status/file/topology/category/Top/limits/basic?range=8195826f5c428fc283510bc67303dd4472d78498..178a32653e369dce2da68dc6b2694e385d484ef1)
* [`category_theory.abelian.opposite`@`8c75ef3517d4106e89fe524e6281d0b0545f47fc`..`a5ff45a1c92c278b03b52459a620cfd9c49ebc80`](https://leanprover-community.github.io/mathlib-port-status/file/category_theory/abelian/opposite?range=8c75ef3517d4106e89fe524e6281d0b0545f47fc..a5ff45a1c92c278b03b52459a620cfd9c49ebc80)
* [`data.mv_polynomial.funext`@`da01792ca4894d4f3a98d06b6c50455e5ed25da3`..`0b89934139d3be96f9dab477f10c20f9f93da580`](https://leanprover-community.github.io/mathlib-port-status/file/data/mv_polynomial/funext?range=da01792ca4894d4f3a98d06b6c50455e5ed25da3..0b89934139d3be96f9dab477f10c20f9f93da580)
* [`category_theory.preadditive.projective`@`f8d8465c3c392a93b9ed226956e26dee00975946`..`3974a774a707e2e06046a14c0eaef4654584fada`](https://leanprover-community.github.io/mathlib-port-status/file/category_theory/preadditive/projective?range=f8d8465c3c392a93b9ed226956e26dee00975946..3974a774a707e2e06046a14c0eaef4654584fada)

### CODE REVIEW

File: Mathlib/CategoryTheory/Abelian/Basic.lean
Comment (eric-wieser) [inline Mathlib/CategoryTheory/Abelian/Basic.lean:1]: Can you update the porting comment for `has_finite_biproducts` to make it clear that this is no longer an instance in mathlib3 either?

File: Mathlib/CategoryTheory/Limits/Preserves/Finite.lean
Comment (eric-wieser) [inline Mathlib/CategoryTheory/Limits/Preserves/Finite.lean:1]: This hasn't forward-ported the docstrings that were added in Mathlib3

Comment (eric-wieser) [reply Mathlib/CategoryTheory/Limits/Preserves/Finite.lean:1]: I also don't see mathlib3's `category_theory.limits.preserves_colimits_of_size.zero.preserves_finite_colimits`

File: Mathlib/CategoryTheory/Preadditive/Projective.lean
Comment (eric-wieser) [inline Mathlib/CategoryTheory/Preadditive/Projective.lean:1]: I'm puzzled by the fact the mathport output for this one hasn't changed!

File: Mathlib/Data/MvPolynomial/Basic.lean [hunk 1]
```diff
@@ -1215,11 +1215,11 @@ theorem eval₂_eq_eval_map (g : σ → S₁) (p : MvPolynomial σ R) : p.eval
     simp only [comp_apply, eval₂_X]
 #align mv_polynomial.eval₂_eq_eval_map MvPolynomial.eval₂_eq_eval_map
 
--- Porting note: new theorem
 -- This probably belongs earlier, but it breaks the fragile proof of `eval₂_eq_eval_map`
 @[simp]
 theorem eval₂_id (p : MvPolynomial σ R) : eval₂ (RingHom.id _) g p = eval g p :=
```
Comment (eric-wieser) [inline Mathlib/Data/MvPolynomial/Basic.lean:1220]: This lemma is right before `eval_eval₂` in Lean 3; can you move it, or add a porting note saying that it moved?

File: Mathlib/Data/MvPolynomial/Funext.lean [hunk 2]
```diff
@@ -4,7 +4,7 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Johan Commelin
 
 ! This file was ported from Lean 3 source module data.mv_polynomial.funext
-! leanprover-community/mathlib commit da01792ca4894d4f3a98d06b6c50455e5ed25da3
+! leanprover-community/mathlib commit 0b89934139d3be96f9dab477f10c20f9f93da580
 ! Please do not edit these lines, except to modify the commit id
 ! if you have ported upstream changes.
 -/
```
Comment (eric-wieser) [inline Mathlib/Data/MvPolynomial/Funext.lean:10]: The imports in this file are in a different order to mathlib3

File: Mathlib/Topology/Category/Top/Limits/Basic.lean
Comment (eric-wieser) [inline Mathlib/Topology/Category/Top/Limits/Basic.lean:1]: I'm a bit concerned about this file; there is no mathport diff to compare to as the file is new. I think the best course of action would be to replace the whole file with the latest mathport output, and then revert back the changes needed to make it build again. I'd prefer this one be removed from the PR as this isn't a simple backport.

Comment (semorrison) [reply Mathlib/Topology/Category/Top/Limits/Basic.lean:1]: All we did was delete the bottom two thirds of the file.

File: Mathlib/CategoryTheory/Abelian/Basic.lean
Comment (semorrison) [reply Mathlib/CategoryTheory/Abelian/Basic.lean:1]: I think simply removing the porting note is best, as it matches the mathlib3 situation exactly now.

File: Mathlib/Data/MvPolynomial/Basic.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Data/MvPolynomial/Basic.lean:1220]: Porting note made clearer.

File: Mathlib/Data/MvPolynomial/Funext.lean (hunk 2 above)
Comment (semorrison) [reply Mathlib/Data/MvPolynomial/Funext.lean:10]: Fixed.

File: Mathlib/Topology/Category/Top/Limits/Basic.lean
Comment (eric-wieser) [reply Mathlib/Topology/Category/Top/Limits/Basic.lean:1]: I don't agree; I just tried to report this from scratch and found a handful of declarations that are in mathlib3 but not in this file.

Comment (semorrison) [reply Mathlib/Topology/Category/Top/Limits/Basic.lean:1]: Okay, copied over the four missing declarations.

File: Mathlib/CategoryTheory/Limits/Preserves/Finite.lean
Comment (semorrison) [reply Mathlib/CategoryTheory/Limits/Preserves/Finite.lean:1]: Okay, I've added that def, forward ported the docstrings, and put in the missing #aligns.

File: Mathlib/CategoryTheory/Abelian/Basic.lean [hunk 3]
```diff
@@ -269,11 +269,6 @@ namespace CategoryTheory.Abelian
 variable {C : Type u} [Category.{v} C] [Abelian C]
 
 /-- An abelian category has finite biproducts. -/
--- Porting note: we would like this to be an instance, and indeed it worked in mathlib3
--- however it triggers https://github.com/leanprover/lean4/issues/2055
--- during the simpNF linter, so for now we turn it on locally in this file.
--- This will likely cause problems in downstream ports, unfortunately.
--- instance (priority := 100)
 theorem hasFiniteBiproducts : HasFiniteBiproducts C :=
```
Comment (eric-wieser) [inline Mathlib/CategoryTheory/Abelian/Basic.lean:272]:
```suggestion
-- Porting note: this should be an instance,
-- but triggers https://github.com/leanprover/lean4/issues/2055
-- We set it as a local instance instead.
-- @[priority 100] instance
theorem hasFiniteBiproducts : HasFiniteBiproducts C :=
```

Note that this comment is present in mathlib3, it seems weird to remove it in mathlib4

File: Mathlib/CategoryTheory/Limits/Preserves/Finite.lean [hunk 4]
```diff
@@ -4,7 +4,7 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Andrew Yang
 
 ! This file was ported from Lean 3 source module category_theory.limits.preserves.finite
-! leanprover-community/mathlib commit 024a4231815538ac739f52d08dd20a55da0d6b23
+! leanprover-community/mathlib commit 3974a774a707e2e06046a14c0eaef4654584fada
```
Comment (eric-wieser) [inline Mathlib/CategoryTheory/Limits/Preserves/Finite.lean:7]: This file seems to still be missing the comments from https://leanprover-community.github.io/mathlib-port-status/file/category_theory/limits/preserves/finite

Comment (eric-wieser) [reply Mathlib/CategoryTheory/Limits/Preserves/Finite.lean:7]: I've pushed the missing comments

File: Mathlib/CategoryTheory/Limits/Preserves/Finite.lean [hunk 5]
```diff
@@ -128,18 +131,21 @@ noncomputable instance (priority := 100) preservesColimitsOfShapeOfPreservesFini
...
 #align category_theory.limits.preserves_colimits_of_shape_of_preserves_finite_colimits CategoryTheory.Limits.preservesColimitsOfShapeOfPreservesFiniteColimits
 
--- Porting note: this is a dangerous instance as it has unbound universe variables.
+-- This is a dangerous instance as it has unbound universe variables.
+/-- If we preserve colimits of some arbitrary size, then we preserve all finite colimits. -/
 noncomputable def PreservesColimitsOfSize.preservesFiniteColimits (F : C ⥤ D)
     [PreservesColimitsOfSize.{w, w₂} F] : PreservesFiniteColimits F where
   preservesFiniteColimits J (sJ : SmallCategory J) fJ := by
     haveI := preservesSmallestColimitsOfPreservesColimits F
     exact preservesColimitsOfShapeOfEquiv (FinCategory.equivAsType J) F
```
Comment (eric-wieser) [inline Mathlib/CategoryTheory/Limits/Preserves/Finite.lean:140]: This one is mising an align

Comment (eric-wieser) [reply Mathlib/CategoryTheory/Limits/Preserves/Finite.lean:140]: probably
```suggestion
noncomputable def PreservesColimitsOfSize.preservesFiniteColimits (F : C ⥤ D)
    [PreservesColimitsOfSize.{w, w₂} F] : PreservesFiniteColimits F where
  preservesFiniteColimits J (sJ : SmallCategory J) fJ := by
    haveI := preservesSmallestColimitsOfPreservesColimits F
    exact preservesColimitsOfShapeOfEquiv (FinCategory.equivAsType J) F
#align category_theory.limits.preserves_colimits_of_size.preserves_finite_colimits CategoryTheory.Limits.PreservesColimitsOfSize.preservesFiniteColimits
```

File: Mathlib/CategoryTheory/Abelian/Basic.lean (hunk 3 above)
Comment (eric-wieser) [reply Mathlib/CategoryTheory/Abelian/Basic.lean:277]: I applied this change.

File: Mathlib/Topology/Category/Top/Limits/Basic.lean [hunk 6]
```diff
@@ -219,3 +219,27 @@ instance forgetPreservesColimitsOfSize :
...
+def isInitialPempty : IsInitial (TopCat.of PEmpty.{u + 1}) :=
+  haveI : ∀ X, Unique (TopCat.of PEmpty.{u + 1} ⟶ X) := fun X =>
+    ⟨⟨⟨fun x => x.elim, by continuity⟩⟩, fun f => by ext ⟨⟩⟩
+  Limits.IsInitial.ofUnique _
+#align Top.is_initial_pempty TopCat.isInitialPempty
+
+/-- The initial object of `Top` is `pempty`. -/
+def initialIsoPempty : ⊥_ TopCat.{u} ≅ TopCat.of PEmpty :=
+  initialIsInitial.uniqueUpToIso isInitialPempty
+#align Top.initial_iso_pempty TopCat.initialIsoPempty
```
Comment (eric-wieser) [inline Mathlib/Topology/Category/Top/Limits/Basic.lean:245]: The comments here have the mathlib3 names for the types

**Comment by eric-wieser** [comment]
bors d+

@semorrison, please check the changes I pushed look reasonable.

In future, could you try not to batch forward-ports together into a single PR? Grouping based on mathlib3 PR would keep the PR a bit smaller, and prevent errors in one forward-port holding up all the others!

**Comment by semorrison** [comment]
Sorry about this PR not being done well. :-( I admit I am finding the zig-zag of "make changes to a mathlib4 port to get things working" / "see if it is possible to backport those changes to mathlib3" / "then have to forward port the additional changes that happen during review of the backport" incredibly frustrating, and am increasingly tempted to just drop the second two steps.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**